docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md

```diff
@@ -0,0 +1,425 @@
+
+Now, fix the typo and run `ruff check` again. This time, you should see no output, which means that `ruff` found no errors in the code. 
+
+This does not, unfortunately, mean that our code is perfect. There are *a lot* of style conventions in the world, and `ruff` can check for many of them. For example, notice that our function does not have a docstring, which most style guides require and which we might well want our linter to flag. 
+
+Let's run `ruff` again, but this time we'll enable all of `ruff`'s checks by passing the `--select ALL` flag:
+```bash
+ruff check --select ALL .
+```
+You should see a minor avalanche of new errors that looks like this:
+```bash
+warning: `one-blank-line-before-class` (D203) and `no-blank-line-before-class` (D211) are incompatible. Ignoring `one-blank-line-before-class`.
+warning: `multi-line-summary-first-line` (D212) and `multi-line-summary-second-line` (D213) are incompatible. Ignoring `multi-line-summary-second-line`.
+main.py:1:1: D100 Missing docstring in public module
+main.py:1:5: ANN201 Missing return type annotation for public function `show_full_name`
+main.py:1:5: D103 Missing docstring in public function
+main.py:1:20: ANN001 Missing type annotation for function argument `first_name`
+main.py:1:44: ANN001 Missing type annotation for function argument `last_name`
+main.py:3:5: T201 `print` found
+Found 8 errors.
+```
+Notice that, among other issues, `ruff` has now noticed that our function (and indeed the `main.py` module itself) does not have a docstring. As an aside, the `warning` lines at the beginning of the output indicate a problem with our use of `ruff`: we have enabled so many lint rules that some of the rules are actually in conflict with one another, so `ruff` has to choose which ones to enforce. 
+
+In general, choosing the appropriate set of lint rules to enforce is a context- and project-dependent decision. As codebases grow and either become more complex or are developed by multiple people, it often makes sense to enforce increasingly strict and extensive linting rules. For our purposes at Arcadia, using the default rules that `ruff` enforces (i.e., using `ruff check` alone) is a great place to start; this will catch many common typos and bugs that will cause errors at runtime (like undefined and unused variables). 
+
+
+### When to format and when to lint
+The short answer is early and often: because formatters and linters are fast and easy to run, it is best to run them frequently and as early as possible in the development process, when errors are easy to fix. Most IDEs can be configured to run formatters automatically each time a file is saved, and linters are often integrate with IDEs as well (this is how VS Code, for example, displays squiggly red lines under undefined variables).
```

Comment:
```suggestion
The short answer is early and often: because formatters and linters are fast and easy to run, it is best to run them frequently and as early as possible in the development process, when errors are easy to fix. Most IDEs can be configured to run formatters automatically each time a file is saved, and linters are often integrated with IDEs as well (this is how VS Code, for example, displays squiggly red lines under undefined variables).
```